### PR TITLE
Add keyring endowment permission

### PR DESCRIPTION
**Pull request by Mrtenz**
This adds a new permission `endowment:keyring` (as described in [SIP-2](https://metamask.github.io/SIPs/SIPS/sip-2#manifest)), with the corresponding caveat.

### CODE REVIEW

File: packages/execution-environments/package.json [hunk 1]
```diff
@@ -32,6 +32,7 @@
     "publish:package": "../../scripts/publish-package.sh"
   },
   "dependencies": {
+    "@metamask/controllers": "^31.1.0",
```
Comment (FrederikBolding) [inline packages/execution-environments/package.json:35]: I dont think this is needed?

File: packages/utils/package.json [hunk 2]
```diff
@@ -56,7 +56,8 @@
     "fast-deep-equal": "^3.1.3",
     "rfdc": "^1.3.0",
     "semver": "^7.3.7",
-    "ses": "^0.15.17"
+    "ses": "^0.15.17",
+    "superstruct": "^0.16.0"
```
Comment (FrederikBolding) [inline packages/utils/package.json:60]: We should make sure this runs in the extension fine.

Related: https://github.com/MetaMask/utils/pull/25

File: packages/controllers/src/snaps/endowments/keyring.test.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        // @ts-expect-error Missing value type.
+        keyringCaveatSpecifications[SnapCaveatType.SnapKeyring].validator?.({
+          type: SnapCaveatType.SnapKeyring,
+        }),
+      ).toThrow('Expected a plain object.');
+
+      expect(() =>
+        keyringCaveatSpecifications[SnapCaveatType.SnapKeyring].validator?.({
+          type: SnapCaveatType.SnapKeyring,
+          value: getNamespace(),
```
Comment (FrederikBolding) [inline packages/controllers/src/snaps/endowments/keyring.test.ts:65]: Why does this throw 🤔

File: packages/utils/src/namespace.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+export const NamespaceStruct = object({
+  /**
+   * A list of supported chains in the namespace.
+   */
+  chains: array(ChainStruct),
+
+  /**
+   * A list of supported RPC methods on the namespace, that a DApp can call.
+   */
+  methods: optional(array(LimitedString)),
```
Comment (FrederikBolding) [inline packages/utils/src/namespace.ts:46]: Why are these limited strings?

File: packages/controllers/src/snaps/endowments/keyring.test.ts (hunk 3 above)
Comment (Mrtenz) [reply packages/controllers/src/snaps/endowments/keyring.test.ts:65]: It must be an object containing the namespace, like
```ts
{
   foo: getNamespace()
}
```

File: packages/utils/src/namespace.ts (hunk 4 above)
Comment (Mrtenz) [reply packages/utils/src/namespace.ts:46]: https://github.com/MetaMask/SIPs/blob/main/assets/sip-2/permission.schema.json#L54-L55

**Review by FrederikBolding — APPROVED** [review]

**Pull request merged**